File: 2022/03/python/main.py
import fileinput
from itertools import batched


class Item(str):
    def priority(self) -> int:
        v = ord(self)

        if 97 <= v <= 122:
            return v - 96
        if 65 <= v <= 90:
            return v - 64 + 26

        raise AssertionError(f"unexpected char {v}")


class RugSack(str):
    def common(self) -> Item:
        left = self[: len(self) // 2]
        right = self[len(self) // 2 :]

        (res,) = set(left).intersection(set(right))
        return Item(res)


class Group(list[RugSack]):
    def batch(self) -> Item:
        (res,) = set(self[0]).intersection(self[1]).intersection(self[2])
        return Item(res)


data: list[RugSack] = []
for line in fileinput.input():
    data.append(RugSack(line[:-1]))

print(f"P1: {sum(i.common().priority() for i in data)}")
print(f"P2: {sum(Group(b).batch().priority() for b in batched(data,3))}")

File: 2022/03/go/main.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strings"
)

type item rune

func (i item) priority() int {
	if int(i) >= 97 && int(i) <= 122 {
		return int(i) - 96
	}

	if int(i) >= 65 && int(i) <= 90 {
		return int(i) - 64 + 26
	}

	panic("this rune should not be in input")
}

type rugsack []item

func (rs rugsack) div() int {
	return len(rs)/2
}

func (rs rugsack) commonItem() item {
	for _, c := range rs[:rs.div()] {
		if strings.ContainsRune(string(rs[rs.div():]), rune(c)) {
			return item(c)
		}
	}

	panic("no common rune found")
}

type group [3]rugsack 


var data []rugsack

func parse(r io.Reader) {
	scanner := bufio.NewScanner(r)

	for scanner.Scan() {
		line := scanner.Text()
		rs := rugsack(line)
		data = append(data, rs)
	}
}

func solve() (res int) {
	for _, rs := range data {
		res += rs.commonItem().priority()
	}

	return
}

func main() {
	parse(os.Stdin)

	P1 := solve()
	P2 := 0

	fmt.Printf("P1: %d \n", P1)
	fmt.Printf("P2: %d \n", P2)
}
